// discord-akairo/dist/src/struct/commands/Flag.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FlagType = void 0;
/**
 * Represents a special return value during command execution or argument parsing.
 */
class Flag {
    constructor(type, data = {}) {
        this.type = type;
        Object.assign(this, data);
    }
    /**
     * Creates a flag that cancels the command.
     */
    static cancel() {
        return new Flag(FlagType.Cancel);
    }
    /**
     * Creates a flag that retries with another input.
     * @param message - Message to handle.
     */
    static retry(message) {
        return new Flag(FlagType.Retry, { message });
    }
    /**
     * Creates a flag that acts as argument cast failure with extra data.
     * @param value - The extra data for the failure.
     */
    static fail(value) {
        return new Flag(FlagType.Fail, { value });
    }
    /**
     * Creates a flag that runs another command with the rest of the arguments.
     * @param command - Command ID.
     * @param ignore - Whether or not to ignore permission checks.
     * @param rest - The rest of the arguments. If this is not set, the argument handler will automatically use the rest of the content.
     */
    static continue(command, ignore = false, rest = null) {
        return new Flag(FlagType.Continue, { command, ignore, rest });
    }
    static is(value, type) {
        return value instanceof Flag && value.type === type;
    }
}
exports.default = Flag;
var FlagType;
(function (FlagType) {
    FlagType["Cancel"] = "cancel";
    FlagType["Retry"] = "retry";
    FlagType["Fail"] = "fail";
    FlagType["Continue"] = "continue";
})(FlagType = exports.FlagType || (exports.FlagType = {}));
